[clang] Enforce 1-based indexing for command line source locations

naveen-seth opened the pull request:
Fixes #139375

Clang expects command line source locations to be provided using 1-based indexing.
Currently, Clang does not reject zero as invalid argument for column or line number, which can cause Clang to crash.

This commit extends validation in `ParsedSourceLocation::FromString` to only accept (unsinged) non-zero integers.

naveen-seth committed:
[clang] Enforce 1-based indexing for command line source locations

Fixes #139375

Clang expects command line source locations to be provided using
1-based indexing.
Currently, Clang does not reject zero as invalid argument for column
or line number, which can cause Clang to crash.

This commit extends validation in `ParsedSourceLocation::FromString`
to only accept (unsinged) non-zero integers.

cor3ntin commented on clang/test/CodeCompletion/crash-if-zero-index.cpp:
```diff
@@ -0,0 +1,6 @@
+
+// Related to #139375
+// Clang uses 1-based indexing for source locations given from the command-line.
+// Verify Clang doesn’t crash when 0 is given as line or column number.
```
Can you add a test using FileCheck that we produce a diagnostic?

cor3ntin commented:
Thanks for working on that.
I think it might better to do that check where `ParsedSourceLocation::FromString` is called, so that we can have a proper
front-end diagnostics for it (search for `err_fe_invalid_code_complete_file` and `OPT_code_completion_at`)

naveen-seth commented:
> Thanks for working on that. I think it might better to do that check where `ParsedSourceLocation::FromString` is called, so that we can have a proper front-end diagnostics for it (search for `err_fe_invalid_code_complete_file` and `OPT_code_completion_at`)

Thank you for the quick review! :)

I added the check in `ParsedSourceLocation::FromString` because it also helps catch the same crash for callers of `ParsedSourceRange::FromString`, which also uses `ParsedSourceLocation::FromString`.

For example, the following would also crash (hitting the same assert as the one this PR fixes)

```bash
touch empty.cpp
clang-refactor local-rename -selection=input.cpp:1:1-1:0 -new-name=test -v input.cpp 
```
<details>
<summary> Full crash log</summary>

```
Error while trying to load a compilation database:
Could not auto-detect compilation database for file "input.cpp"
No compilation database found in /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-debug-env/debugging-139375 or any parent directory
fixed-compilation-database: Error while opening fixed database: No such file or directory
json-compilation-database: Error while opening JSON database: No such file or directory
Running without flags.
clang-refactor: /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/clang/lib/Basic/SourceManager.cpp:1666: SourceLocation clang::SourceManager::translateLineCol(FileID, unsigned int, unsigned int) const: Assertion `Line && Col && "Line and column should start from 1!"' failed.
 #0 0x0000562e7ef658cd llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/llvm/lib/Support/Unix/Signals.inc:804:11
 #1 0x0000562e7ef65e0b PrintStackTraceSignalHandler(void*) /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/llvm/lib/Support/Unix/Signals.inc:888:1
 #2 0x0000562e7ef63f2f llvm::sys::RunSignalHandlers() /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/llvm/lib/Support/Signals.cpp:105:5
 #3 0x0000562e7ef664e9 SignalHandler(int, siginfo_t*, void*) /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/llvm/lib/Support/Unix/Signals.inc:418:7
 #4 0x00007f6311a5e250 (/lib/x86_64-linux-gnu/libc.so.6+0x45250)
 #5 0x00007f6311abcf1c pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0xa3f1c)
 #6 0x00007f6311a5e19e raise (/lib/x86_64-linux-gnu/libc.so.6+0x4519e)
 #7 0x00007f6311a41902 abort (/lib/x86_64-linux-gnu/libc.so.6+0x28902)
 #8 0x00007f6311a4181e (/lib/x86_64-linux-gnu/libc.so.6+0x2881e)
 #9 0x00007f6311a547c7 (/lib/x86_64-linux-gnu/libc.so.6+0x3b7c7)
#10 0x0000562e7fbee816 clang::SourceManager::translateLineCol(clang::FileID, unsigned int, unsigned int) const /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/clang/lib/Basic/SourceManager.cpp:1668:11
#11 0x0000562e7ed5dc34 (anonymous namespace)::SourceRangeSelectionArgument::forAllRanges(clang::SourceManager const&, llvm::function_ref<void (clang::SourceRange)>) /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/clang/tools/clang-refactor/ClangRefactor.cpp:131:12
#12 0x0000562e7ed5eb2b (anonymous namespace)::ClangRefactorTool::callback(clang::ASTContext&) /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/clang/tools/clang-refactor/ClangRefactor.cpp:418:11
#13 0x0000562e7ed5e840 (anonymous namespace)::ClangRefactorTool::getFrontendActionFactory()::'lambda'(clang::ASTContext&)::operator()(clang::ASTContext&) const /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/clang/tools/clang-refactor/ClangRefactor.cpp:475:35
#14 0x0000562e7ed5e80d void std::__invoke_impl<void, (anonymous namespace)::ClangRefactorTool::getFrontendActionFactory()::'lambda'(clang::ASTContext&)&, clang::ASTContext&>(std::__invoke_other, (anonymous namespace)::ClangRefactorTool::getFrontendActionFactory()::'lambda'(clang::ASTContext&)&, clang::ASTContext&) /usr/lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/invoke.h:61:7
#15 0x0000562e7ed5e7ad std::enable_if<is_invocable_r_v<void, (anonymous namespace)::ClangRefactorTool::getFrontendActionFactory()::'lambda'(clang::ASTContext&)&, clang::ASTContext&>, void>::type std::__invoke_r<void, (anonymous namespace)::ClangRefactorTool::getFrontendActionFactory()::'lambda'(clang::ASTContext&)&, clang::ASTContext&>((anonymous namespace)::ClangRefactorTool::getFrontendActionFactory()::'lambda'(clang::ASTContext&)&, clang::ASTContext&) /usr/lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/invoke.h:117:5
#16 0x0000562e7ed5e6f5 std::_Function_handler<void (clang::ASTContext&), (anonymous namespace)::ClangRefactorTool::getFrontendActionFactory()::'lambda'(clang::ASTContext&)>::_M_invoke(std::_Any_data const&, clang::ASTContext&) /usr/lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/std_function.h:290:2
#17 0x0000562e7ed774c6 std::function<void (clang::ASTContext&)>::operator()(clang::ASTContext&) const /usr/lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/std_function.h:591:2
#18 0x0000562e7ed5f3b1 (anonymous namespace)::ClangRefactorTool::getFrontendActionFactory()::ToolASTConsumer::HandleTranslationUnit(clang::ASTContext&) /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/clang/tools/clang-refactor/ClangRefactor.cpp:441:7
#19 0x0000562e80984ffb clang::ParseAST(clang::Sema&, bool, bool) /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/clang/lib/Parse/ParseAST.cpp:191:12
#20 0x0000562e800193d9 clang::ASTFrontendAction::ExecuteAction() /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/clang/lib/Frontend/FrontendAction.cpp:1345:1
#21 0x0000562e80018e26 clang::FrontendAction::Execute() /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/clang/lib/Frontend/FrontendAction.cpp:1229:7
#22 0x0000562e7ff2f0b3 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/clang/lib/Frontend/CompilerInstance.cpp:1057:23
#23 0x0000562e8044f193 clang::tooling::FrontendActionFactory::runInvocation(std::shared_ptr<clang::CompilerInvocation>, clang::FileManager*, std::shared_ptr<clang::PCHContainerOperations>, clang::DiagnosticConsumer*) /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/clang/lib/Tooling/Tooling.cpp:466:14
#24 0x0000562e8044f05b clang::tooling::ToolInvocation::runInvocation(char const*, clang::driver::Compilation*, std::shared_ptr<clang::CompilerInvocation>, std::shared_ptr<clang::PCHContainerOperations>) /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/clang/lib/Tooling/Tooling.cpp:441:18
#25 0x0000562e8044e217 clang::tooling::ToolInvocation::run() /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/clang/lib/Tooling/Tooling.cpp:426:3
#26 0x0000562e80450382 clang::tooling::ClangTool::run(clang::tooling::ToolAction*) /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/clang/lib/Tooling/Tooling.cpp:624:11
#27 0x0000562e7ed595ea main /home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/clang/tools/clang-refactor/ClangRefactor.cpp:636:38
#28 0x00007f6311a433b8 (/lib/x86_64-linux-gnu/libc.so.6+0x2a3b8)
#29 0x00007f6311a4347b __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2a47b)
#30 0x0000562e7ed591e5 _start (/home/nav/.distrobox/ubuntu-24-10-compiler-dev/llvm-project-dev/build/bin/clang-refactor+0x23741e5)
Aborted (core dumped)
```

</details>

(And similar invalid inputs like `<file>:<line>:-1` are caught in `ParsedSourceLocation::FromString`)


I could move the check up to the locations where `ParsedSourceLocation::FromString` and `ParsedSourceRange::FromString` are called, but I believe it should be possible to provide a proper front-end diagnostic without moving the check.
For context, this PR currently changes the behavior of the crash in #139375 to also omit the `diag::err_drv_invalid_value` diagnostic, but I could adjust it to hint at the right format.

What do you think is best here?

naveen-seth committed:
Test for diagnostic using FileCheck

AaronBallman reviewed (commented):
Thank you for the fix! The change should come with a release note in `clang/docs/ReleaseNotes.rst` so users know about the improvement.

Otherwise, I think this is going in a reasonable direction.

naveen-seth committed:
Add tests for clang-refactor

naveen-seth committed:
Ensure 1-based indexing for end of ParsedSourceRange as well

naveen-seth committed:
Add release notes

naveen-seth committed:
Improve diagnostics

MaskRay commented on clang/include/clang/Frontend/CommandLineSourceLoc.h:
```diff
@@ -89,8 +92,13 @@ struct ParsedSourceRange {
         RangeSplit.first = Str;
-      } else
+      } else {
+        // Column and line numbers are 1-based
```
Ensure that full sentences in commends end with `.`

Remove brace for this single-line-simple-body case per https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

MaskRay commented on clang/test/CodeCompletion/crash-command-line-source-loc-zero.cpp:
```diff
@@ -0,0 +1,12 @@
+// Related to #139375
```
I don't why some regression tests are named `crash-*`, but I would not name this `crash-*`.

I'd just name this source-loc-zero.cpp and change this comment to `// Regression test for #139375`

MaskRay commented on clang/test/Refactor/crash-command-line-source-loc-zero.cpp:
```diff
@@ -0,0 +1,19 @@
+// RUN: not clang-refactor local-rename -selection=%s:1:1-1:0 -new-name=test %s 2>&1 \
+// RUN:     | FileCheck -check-prefix=CHECK-DIAG %s
+
+// CHECK-DIAG: error: '-selection' option must be specified using <file>:<line>:<column>
```
If the diagnostic is on one line, use a single CHECK.

Multiple CHECK accepts output that we don't expect.

We don't rigidly require line wrapping for tests.

MaskRay reviewed (approved):
LGTM with a few nits. Please give other reviewers some time to chime in. Click the "Re-request review" circle icon beside reviewer usernames after you make an update.

naveen-seth committed:
Address comments

AaronBallman commented on clang/test/Refactor/source-loc-zero.cpp:
```diff
@@ -0,0 +1,17 @@
+// RUN: not clang-refactor local-rename -selection=%s:1:1-1:0 -new-name=test %s 2>&1 \
+// RUN:     | FileCheck -check-prefix=CHECK-DIAG %s
+
+// CHECK-DIAG: error: '-selection' option must be specified using <file>:<line>:<column> or <file>:<line>:<column>-<line>:<column> format, where <line> and <column> are integers greater than zero.
```
Can you add a newline to the end of the file?

AaronBallman reviewed (approved):
Other than a nit, LGTM!

naveen-seth committed:
Add newline

naveen-seth commented:
Thank you for reviewing!
(And apologies for the formatting and coding-standard issues on my end!)

AaronBallman commented:
> Thank you for reviewing! (And apologies for the formatting and coding-standard issues on my end.)

No need to apologize, that sort of thing comes up! :-)

Do you need one of us to merge the changes on your behalf?
